Annotation processor can't handle more than 5 library modules

<!--
Please fill in the below fields with some data to help us best diagnose the issue.
The more specific you are, the better! You can help a lot by not making us ask these questions.
Feel free to remove any irrelevant parts that you know are not related to the issue.
Any HTML comment like this will be stripped when rendering markdown, no need to delete them.
-->


<!-- What version of Glide you're running, for example: 3.7.1 | 3.8.0-SNAPSHOT | 4.0.0-SNAPSHOT
It's essentially the version number from your build.gradle: `dependencies { compile '...:x.y.z' }` -->
**Glide Version**: 4.12.0

<!-- Do you use any integration library, like OkHttp3 or Volley? For example:
Fails to display with stock networking, but works with okhttp3-1.4.0 -->
**Integration libraries**: none / all

<!-- What devices you managed to get the issue to come up on? For example:
fails on Galaxy S4/GT-I9500 4.4.2, works fine on Nexus 6P 5.1 and Genymotion Nexus 5 5.0.1 -->
**Device/Android Version**: all

**Build system**: Gradle 5.6.4 on Windows 10

<!-- Share the details of your issue in prose, detailing actual and expected behavior. It also helps if you give some info **why** you are trying to do something as opposed to **what** is not working. -->
**Issue details / Repro steps / Use case background**:
Having 5-6 library modules in the project (depending on package and class names) stops AppGlideModule from compiling.

1. Get this project: https://github.com/TWiStErRob/repros/tree/master/glide/apt_many_libs
1. `gradlew assembleDebug --stacktrace` -> fails

Expected:
same behavior as when there are less modules, e.g.
1. Delete `LibModule6.java`, or comment it out, or remove `@GlideModule`.
1. `gradlew assembleDebug --stacktrace` -> all good

Actual: see stack trace.

<!-- How do you use Glide?
Make sure you include everything as is in your app's code:
Changing a single method parameter can yield totally different results.
Please clarify any magic variables that appear in the code, for example: "// `this` is a Fragment"
-->
**Glide load line / `GlideModule` (if any) / list Adapter code (if any)**:
```java
package com.example.application.pack.age;
@GlideModule public final class AppModule extends AppGlideModule { }
@GlideModule public class LibModule1 extends LibraryGlideModule { }
@GlideModule public class LibModule2 extends LibraryGlideModule { }
@GlideModule public class LibModule3 extends LibraryGlideModule { }
@GlideModule public class LibModule4 extends LibraryGlideModule { }
@GlideModule public class LibModule5 extends LibraryGlideModule { }
@GlideModule public class LibModule6 extends LibraryGlideModule { }
```

<!-- How does your app look like?
We're most interested in the layout attributes and the hierarchy around the ImageView -->
**Layout XML**: N/A

<!--
What is the error message that you got in the log?
You can find some help on diagnosing issues here: https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling
-->
**Stack trace / LogCat**:
<details>
  <summary>

Execution failed for task ':compileDebugJavaWithJavac'.
  > java.lang.RuntimeException: java.io.FileNotFoundException: 
 > P:\projects\bugs\glide\apt_many_libs\build\generated\ap_generated_sources\debug\out\com\bumptech\glide\annotation\compiler\GlideIndexer_GlideModule_com_example_application_pack_age_LibModule1_com_example_application_pack_age_LibModule2_com_example_application_pack_age_LibModule3_com_example_application_pack_age_LibModule4_com_example_application_pack_age_LibModule5_com_example_application_pack_age_LibModule6.java (The filename, directory name, or volume label syntax is incorrect)
  **Click to expand...**
  </summary>

  ```ruby
  > Task :compileDebugJavaWithJavac FAILED

  FAILURE: Build failed with an exception.

  * What went wrong:
  Execution failed for task ':compileDebugJavaWithJavac'.
  > java.lang.RuntimeException: java.io.FileNotFoundException: P:\projects\bugs\glide\apt_many_libs\build\generated\ap_generated_sources\debug\out\com\bumptech\glide\annotation\compiler\GlideIndexer_GlideModule_com_example_application_pack_age_LibModule1_com_example_application_pack_age_LibModule
  2_com_example_application_pack_age_LibModule3_com_example_application_pack_age_LibModule4_com_example_application_pack_age_LibModule5_com_example_application_pack_age_LibModule6.java (The filename, directory name, or volume label syntax is incorrect)

  * Try:
  Run with --info or --debug option to get more log output. Run with --scan to get full insights.

  * Exception is:
  org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileDebugJavaWithJavac'.
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:166)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:163)
          at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:191)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:156)
          at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:62)
          at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:108)
          at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
          at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
          at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
          at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
          at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
          at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
          at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
          at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
          at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
          at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
          at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
          at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
          at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
          at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
          at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
          at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
          at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
          at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
          at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
          at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
          at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
          at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
          at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
          at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
          at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
          at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
          at java.lang.Thread.run(Thread.java:748)
  Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.io.FileNotFoundException: P:\projects\bugs\glide\apt_many_libs\build\generated\ap_generated_sources\debug\out\com\bumptech\glide\annotation\compiler\GlideIndexer_GlideModule_com_example_application_pack_age_LibModule1_com_e
  xample_application_pack_age_LibModule2_com_example_application_pack_age_LibModule3_com_example_application_pack_age_LibModule4_com_example_application_pack_age_LibModule5_com_example_application_pack_age_LibModule6.java (The filename, directory name, or volume label syntax is incorrect)
          at com.sun.tools.javac.main.Main.compile(Main.java:553)
          at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
          at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
          at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:93)
          at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
          at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
          at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
          at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:100)
          at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:52)
          at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:38)
          at org.gradle.api.internal.tasks.compile.AnnotationProcessorDiscoveringCompiler.execute(AnnotationProcessorDiscoveringCompiler.java:51)
          at org.gradle.api.internal.tasks.compile.AnnotationProcessorDiscoveringCompiler.execute(AnnotationProcessorDiscoveringCompiler.java:37)
          at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:39)
          at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilerFactory$2.execute(IncrementalCompilerFactory.java:101)
          at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilerFactory$2.execute(IncrementalCompilerFactory.java:97)
          at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:60)
          at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:44)
          at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:59)
          at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:51)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
          at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
          at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler.execute(CompileJavaBuildOperationReportingCompiler.java:51)
          at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:158)
          at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:126)
          at com.android.build.gradle.tasks.AndroidJavaCompile.compile(AndroidJavaCompile.kt:237)
          at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
          at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
          at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
          at java.lang.reflect.Method.invoke(Method.java:498)
          at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
          at org.gradle.api.internal.project.taskfactory.IncrementalTaskInputsTaskAction.doExecute(IncrementalTaskInputsTaskAction.java:47)
          at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
          at org.gradle.api.internal.project.taskfactory.AbstractIncrementalTaskAction.execute(AbstractIncrementalTaskAction.java:25)
          at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:476)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
          at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
          at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:461)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:444)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:93)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:237)
          at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$0(ExecuteStep.java:32)
          at java.util.Optional.map(Optional.java:215)
          at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:32)
          at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
          at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:58)
          at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
          at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
          at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:33)
          at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:39)
          at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
          at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
          at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:35)
          at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
          at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:45)
          at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:31)
          at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:208)
          at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:70)
          at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:45)
          at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
          at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:43)
          at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:32)
          at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
          at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
          at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
          at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:89)
          at java.util.Optional.map(Optional.java:215)
          at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
          at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
          at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
          at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
          at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
          at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
          at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:90)
          at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:48)
          at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:69)
          at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:47)
          at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
          at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:140)
          ... 37 more
  Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: P:\projects\bugs\glide\apt_many_libs\build\generated\ap_generated_sources\debug\out\com\bumptech\glide\annotation\compiler\GlideIndexer_GlideModule_com_example_application_pack_age_LibModule1_com_example_application_pack_age_
  LibModule2_com_example_application_pack_age_LibModule3_com_example_application_pack_age_LibModule4_com_example_application_pack_age_LibModule5_com_example_application_pack_age_LibModule6.java (The filename, directory name, or volume label syntax is incorrect)
          at com.bumptech.glide.annotation.compiler.ProcessorUtil.writeClass(ProcessorUtil.java:133)
          at com.bumptech.glide.annotation.compiler.ProcessorUtil.writeIndexer(ProcessorUtil.java:122)
          at com.bumptech.glide.annotation.compiler.LibraryModuleProcessor.processModules(LibraryModuleProcessor.java:47)
          at com.bumptech.glide.annotation.compiler.GlideAnnotationProcessor.process(GlideAnnotationProcessor.java:112)
          at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
          at org.gradle.api.internal.tasks.compile.processing.AggregatingProcessor.process(AggregatingProcessor.java:50)
          at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
          at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.access$401(TimeTrackingProcessor.java:37)
          at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:99)
          at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:96)
          at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.track(TimeTrackingProcessor.java:117)
          at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.process(TimeTrackingProcessor.java:96)
          at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
          at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
          at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
          at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
          at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
          at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
          at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
          at com.sun.tools.javac.main.Main.compile(Main.java:523)
          ... 125 more
  Caused by: java.io.FileNotFoundException: P:\projects\bugs\glide\apt_many_libs\build\generated\ap_generated_sources\debug\out\com\bumptech\glide\annotation\compiler\GlideIndexer_GlideModule_com_example_application_pack_age_LibModule1_com_example_application_pack_age_LibModule2_com_example_appli
  cation_pack_age_LibModule3_com_example_application_pack_age_LibModule4_com_example_application_pack_age_LibModule5_com_example_application_pack_age_LibModule6.java (The filename, directory name, or volume label syntax is incorrect)
          at java.io.FileOutputStream.open0(Native Method)
          at java.io.FileOutputStream.open(FileOutputStream.java:270)
          at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
          at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
          at com.sun.tools.javac.file.RegularFileObject.openWriter(RegularFileObject.java:136)
          at com.sun.tools.javac.processing.JavacFiler$FilerWriter.<init>(JavacFiler.java:286)
          at com.sun.tools.javac.processing.JavacFiler$FilerOutputFileObject.openWriter(JavacFiler.java:136)
          at com.bumptech.glide.repackaged.com.squareup.javapoet.JavaFile.writeTo(JavaFile.java:116)
          at com.bumptech.glide.annotation.compiler.ProcessorUtil.writeClass(ProcessorUtil.java:131)
          ... 144 more


  * Get more help at https://help.gradle.org

  BUILD FAILED in 1s
  ```
</details>

<!-- Bonus points if you attach a relevant screenshot, screen recording or a small demo project -->

--- a/annotation/compiler/src/main/java/com/bumptech/glide/annotation/compiler/IndexerGenerator.java
+++ b/annotation/compiler/src/main/java/com/bumptech/glide/annotation/compiler/IndexerGenerator.java
@@ -8,6 +8,7 @@ import com.squareup.javapoet.TypeSpec;
 import java.lang.annotation.Annotation;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.UUID;
 import javax.lang.model.element.Modifier;
 import javax.lang.model.element.TypeElement;
 
@@ -46,6 +47,7 @@ import javax.lang.model.element.TypeElement;
  */
 final class IndexerGenerator {
   private static final String INDEXER_NAME_PREFIX = "GlideIndexer_";
+  private static final int MAXIMUM_FILE_NAME_LENGTH = 255;
   private final ProcessorUtil processorUtil;
 
   IndexerGenerator(ProcessorUtil processorUtil) {
@@ -79,7 +81,7 @@ final class IndexerGenerator {
     }
   }
 
-  private static TypeSpec generate(
+  private TypeSpec generate(
       List<TypeElement> libraryModules, Class<? extends Annotation> annotation) {
     AnnotationSpec.Builder annotationBuilder = AnnotationSpec.builder(Index.class);
 
@@ -88,15 +90,25 @@ final class IndexerGenerator {
       annotationBuilder.addMember(value, "$S", ClassName.get(childModule).toString());
     }
 
-    StringBuilder indexerName =
+    StringBuilder indexerNameBuilder =
         new StringBuilder(INDEXER_NAME_PREFIX + annotation.getSimpleName() + "_");
     for (TypeElement element : libraryModules) {
-      indexerName.append(element.getQualifiedName().toString().replace(".", "_"));
-      indexerName.append("_");
+      indexerNameBuilder.append(element.getQualifiedName().toString().replace(".", "_"));
+      indexerNameBuilder.append("_");
     }
-    indexerName = new StringBuilder(indexerName.substring(0, indexerName.length() - 1));
+    indexerNameBuilder =
+        new StringBuilder(indexerNameBuilder.substring(0, indexerNameBuilder.length() - 1));
+    String indexerName = indexerNameBuilder.toString();
+    // If the indexer name has too many packages/modules, it can exceed the file name length
+    // allowed by the file system, which can break compilation. To avoid that, fall back to a
+    // deterministic UUID.
+    if (indexerName.length() >= (MAXIMUM_FILE_NAME_LENGTH - INDEXER_NAME_PREFIX.length())) {
+      indexerName =
+          INDEXER_NAME_PREFIX
+              + UUID.nameUUIDFromBytes(indexerName.getBytes()).toString().replace("-", "_");
+    }
 
-    return TypeSpec.classBuilder(indexerName.toString())
+    return TypeSpec.classBuilder(indexerName)
         .addAnnotation(annotationBuilder.build())
         .addModifiers(Modifier.PUBLIC)
         .build();

--- a/annotation/compiler/test/src/test/java/com/bumptech/glide/annotation/compiler/OverlyLongFileNameTest.java
+++ b/annotation/compiler/test/src/test/java/com/bumptech/glide/annotation/compiler/OverlyLongFileNameTest.java
@@ -1,0 +1,71 @@
+package com.bumptech.glide.annotation.compiler;
+
+import static com.google.testing.compile.Compiler.javac;
+
+import com.bumptech.glide.annotation.compiler.test.CompilationProvider;
+import com.google.testing.compile.Compilation;
+import com.google.testing.compile.CompilationSubject;
+import com.google.testing.compile.JavaFileObjects;
+import java.io.File;
+import java.io.IOException;
+import javax.tools.JavaFileObject;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Makes sure that we can handle indexers based on long package or file names, or many modules.
+ *
+ * <p>See #4106.
+ */
+@RunWith(JUnit4.class)
+public class OverlyLongFileNameTest implements CompilationProvider {
+  @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
+  private Compilation compilation;
+  private static final String FILE_NAME_LONGER_THAN_255_CHARS =
+      "SomeReallyReallyRidiculouslyLongFileNameOrPackageNameIGuessThatExceedsTwoHundredAndFiftyFive"
+          + "CharactersThoughThatsOnlyAroundOneHundredCharactersWhichMeansINeedToKeepTypingToGetTo"
+          + "TwoHundredAndFiftyFiveSomehowThankfullyOnlyLikeFiftyToGoNowMaybeButNotQuiteYet"
+          + "SomewhereAroundNowIsProbablyGood";
+
+  @Before
+  public void setUp() {
+    compilation =
+        javac()
+            .withProcessors(new GlideAnnotationProcessor())
+            .compile(
+                JavaFileObjects.forSourceLines(
+                    FILE_NAME_LONGER_THAN_255_CHARS,
+                    "package com.bumptech.glide.test;",
+                    "import com.bumptech.glide.annotation.GlideModule;",
+                    "import com.bumptech.glide.module.LibraryGlideModule;",
+                    "@GlideModule",
+                    "public final class "
+                        + FILE_NAME_LONGER_THAN_255_CHARS
+                        + " extends LibraryGlideModule {}"));
+  }
+
+  @Test
+  public void compilingLongClassAndOrPackageNameShouldSucceed() throws IOException {
+    CompilationSubject.assertThat(compilation).succeededWithoutWarnings();
+    for (JavaFileObject file : compilation.generatedFiles()) {
+      temporaryFolder.create();
+      String actualFileName = new File(file.getName()).getName();
+      if (!actualFileName.startsWith(FILE_NAME_LONGER_THAN_255_CHARS)) {
+        try {
+          temporaryFolder.newFile(actualFileName).createNewFile();
+        } catch (IOException e) {
+          throw new RuntimeException("Failed to create: " + actualFileName, e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public Compilation getCompilation() {
+    return compilation;
+  }
+}

